settings_notifications: Migrate module to typescript

evykassirer opened the pull request:
built off of #31718, merge that first

I got stuck on a few things, so I left `TODO(evy)` comments there and am putting this up as a draft.

evykassirer commented on web/src/settings_notifications.ts:
```diff
@@ -11,19 +12,22 @@ import * as message_notifications from "./message_notifications";
+import * as util from "./util";
 
-export const user_settings_panel = {};
+export let user_settings_panel: SettingsPanel | undefined;
```
maybe I should do a prep commit to initialize this to `undefined`

evykassirer commented on web/src/settings_notifications.ts:
```diff
@@ -137,7 +154,19 @@ function stream_notification_setting_changed(e) {
     const $status_element = $(e.target).closest(".subsection-parent").find(".alert-notification");
-    const setting = e.target.name;
+    assert(e.target instanceof HTMLInputElement);
+    // TODO(evy): Is this the best we can do here?
+    assert(
+        [
+            "desktop_notifications",
+            "audible_notifications",
+            "email_notifications",
+            "wildcard_mentions_notify",
+            "push_notifications",
+        ].includes(e.target.name),
+    );
+    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+    const setting = e.target.name as keyof StreamSpecificNotificationSettings;
```
You'd think by now I'd remember that my typescript questions are almost always answered by Zod, thanks!

evykassirer commented on web/src/settings_notifications.ts:
```diff
@@ -255,9 +291,16 @@ export function set_up(settings_panel) {
+        assert(setting_value !== undefined && setting_value !== null);
 
         if (
+            // TODO(evy): this is causing an issue because `pm_mention_notification_settings`
+            // has a very specific type and so any valid call to `includes` would be true.
+            // Not sure what to do here.
```
I don't understand how to use Zod to fix this issue, can you elaborate?

timabbott commented on web/src/settings_notifications.ts:
```diff
@@ -9,21 +11,30 @@ import * as confirm_dialog from "./confirm_dialog";
+    pm_notification_settings_schema,
+    user_settings,
+    user_settings_schema,
+} from "./user_settings";
```
Should we just do a normal import of `user_settings` here?

evykassirer replied:
That seems reasonable.

evykassirer replied:
It's causing some line overflow that's not the nicest, I'll push it and you can let me know what you think

timabbott replied:
Line overflow is fine.

timabbott commented on web/src/settings_notifications.ts:
```diff
@@ -149,16 +173,19 @@ function stream_notification_setting_changed(e) {
     const settings_object = settings_panel.settings_object;
-    const $notification_sound_elem = $(settings_panel.notification_sound_elem);
+    assert(settings_panel.notification_sound_elem !== null);
+    const $notification_sound_elem = $<HTMLVideoElement>(settings_panel.notification_sound_elem);
```
Is this a video element? Probably deserves a comment, since it's definitely an element for a sound file, if there's some reason we have to use the video type here.

evykassirer replied:
Oh oops, I can just add `HTMLAudioElement` to the definition of `play_audio`

timabbott commented on web/src/settings_notifications.ts:
```diff
@@ -343,20 +379,21 @@ export function update_page(settings_panel) {
                 break;
             }
             default: {
+                const valid_setting = realm_default_settings_schema.keyof().parse(setting);
```
Why is this schema a valid way to parse this? `realm_default_settings_schema` isn't obviously related to notification settings.

evykassirer replied:
Oh thanks for catching this, I had gotten confused but realized I should update this type to be more clear about what kinds of settings we're working with:



```js
export type SettingsPanel = {
    container: string;
    notification_sound_elem: string | null;
} & ({
    settings_object: UserSettings;
    for_realm_settings: false;
} | {
    settings_object: RealmDefaultSettings;
    for_realm_settings: true;
});
```

andersk commented on web/src/settings_notifications.ts:
```diff
@@ -371,19 +409,22 @@ export function update_muted_stream_state(sub) {
+    user_settings_panel = {
+        container: "#user-notification-settings",
+        settings_object: user_settings.user_settings,
+        notification_sound_elem: "#user-notification-sound-audio",
```
```suggestion
        notification_sound_elem: "audio#user-notification-sound-audio",
```

to justify the `HTMLAudioElement` annotation. Also for the corresponding string in `settings_realm_user_settings_defaults.js`.

andersk commented on web/src/ui_util.ts:
```diff
@@ -169,7 +169,7 @@ export function parse_html(html: string): DocumentFragment {
  * https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/play for more details.
  */
-export async function play_audio(elem: HTMLVideoElement): Promise<void> {
+export async function play_audio(elem: HTMLVideoElement | HTMLAudioElement): Promise<void> {
```
```suggestion
export async function play_audio(elem: HTMLAudioElement): Promise<void> {
```

`HTMLVideoElement` was just wrong here, this is only called with `HTMLAudioElement`.

evykassirer replied:
Ah okay, I'll fix that in a prep commit, thanks!

andersk commented on web/src/settings_notifications.ts:
```diff
@@ -77,7 +87,11 @@ function update_desktop_icon_count_display(settings_panel) {
+export function set_notification_batching_ui(
+    $container: JQuery,
+    setting_seconds: number,
+    force_custom?: boolean,
```
```suggestion
    force_custom = false,
```

to avoid the `=== true` game.

andersk commented on web/src/settings_notifications.ts:
```diff
@@ -149,16 +169,19 @@ function stream_notification_setting_changed(e) {
     const for_realm_settings = settings_panel.for_realm_settings;
-    const $notification_sound_dropdown = $container.find(".setting_notification_sound");
+    const $notification_sound_dropdown = $container.find<HTMLSelectElement & {type: "select-one"}>(
+        ".setting_notification_sound",
```
```suggestion
        "select:not([multiple]).setting_notification_sound",
```

to justify the annotation.

andersk commented on web/src/settings_notifications.ts:
```diff
@@ -137,10 +154,13 @@ function stream_notification_setting_changed(e) {
 
     const $status_element = $(e.target).closest(".subsection-parent").find(".alert-notification");
-    const setting = e.target.name;
+    assert(e.target instanceof HTMLInputElement);
```
This function only uses `e` for its `target`, so it could instead be passed the `HTMLInputElement` instead of the `JQuery.ChangeEvent`. The selector is an event handler for the selector `"input, select"` and `<input>` has no children, so there’s no difference here between `.target` and `.currentTarget`, a.k.a. the handler’s `this`, which could be typed `HTMLInputElement | HTMLSelectElement`. So we’d still need to check, but it’d make sense what we’re checking for.

(In general, I think almost all of our uses of `.target` should be `.currentTarget`/`this`, which has clearer semantics and less type confusion.)

evykassirer replied:
That makes sense. Added a prep commit for this, thanks!

evykassirer commented:
Addressed comments, and put the selector specifying in their own commits just in case of bugs, but I tested it manually and the new selectors work fine as far as I can tell.

andersk commented on web/src/settings_notifications.ts:
```diff
@@ -216,88 +237,101 @@ export function set_up(settings_panel) {
-                    $container,
-                    settings_object.email_notifications_batching_period_seconds,
-                    true,
+    $container
+        .find(".notification-settings-form")
+        .on("change", "input, select", function (this: HTMLElement, e) {
```
We can prevent Prettier from indenting this whole function an extra level by extracting a variable:

```js
    const $notification_settings_form = $container.find(".notification-settings-form");
    $notification_settings_form.on("change", "input, select", function (this: HTMLElement, e) {
```

andersk commented on web/src/settings_preferences.ts:
```diff
@@ -20,20 +20,26 @@ import * as overlays from "./overlays";
 import {page_params} from "./page_params";
 import type {RealmDefaultSettings} from "./realm_user_settings_defaults";
 import * as settings_components from "./settings_components";
-import type {GenericUserSettings} from "./settings_config";
```
We can delete the unused `GenericUserSettings` from `settings_config.ts`, too.

andersk commented on web/src/settings_notifications.ts:
```diff
@@ -11,19 +13,23 @@ import * as message_notifications from "./message_notifications";
 import * as stream_settings_api from "./stream_settings_api";
 import * as stream_settings_data from "./stream_settings_data";
+import {stream_specific_notification_settings_schema} from "./stream_types";
 import * as sub_store from "./sub_store";
+import type {StreamSubscription} from "./sub_store";
 import * as ui_util from "./ui_util";
 import * as unread_ui from "./unread_ui";
-import {user_settings} from "./user_settings";
+import * as user_settings from "./user_settings";
```
I’m not convinced this wildcard import change was a successful experiment. This module’s export names were designed to make sense in an unqualified context, and `user_settings.user_settings[setting_name]` is quite a mouthful. @timabbott thoughts?

evykassirer replied:
I'll wait for @timabbott to reply to this before making the other changes you suggested

timabbott replied:
Yeah this seems like a mistake; I did not expect a `user_settings.user_settings` to be a thing.

evykassirer replied:
Switched back to importing things individually from `user_settings`

andersk reviewed (commented):
Looks good with minor formatting nits.

timabbott commented:
Merged, thanks @evykassirer and @andersk!
